daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -401,15 +378,85 @@ object ImmArray extends ImmArrayInstances {
 
   private[ImmArray] final class IACanBuildFrom[A]
       extends CanBuildFrom[ImmArray[_], A, ImmArray[A]] {
-    override def apply(from: ImmArray[_]) = newBuilder[A]
-    override def apply() = newBuilder[A]
+    override def apply(from: ImmArray[_]) = newBuilder[A]()
+    override def apply() = newBuilder[A]()
   }
 
   implicit def `ImmArray canBuildFrom`[A]: CanBuildFrom[ImmArray[_], A, ImmArray[A]] =
     new IACanBuildFrom
 
-  def newBuilder[A]: mutable.Builder[A, ImmArray[A]] =
-    mutable.ArraySeq.newBuilder[A].mapResult(ImmArray.unsafeFromArraySeq)
+  private[this] def copyArraySeq[A](
+      src: mutable.ArraySeq[A],
+      srcStart: Int,
+      dst: mutable.ArraySeq[A],
+      dstStart: Int,
+      len: Int,
+  ): Unit = {
+    var i = srcStart
+    var j = dstStart
+    val end = srcStart + len
+    while (i < end) {
+      dst(j) = src(i)
+      i += 1
+      j += 1
+    }
+  }
+
+  def newBuilder[A](initialCapacity: Int = 4): mutable.Builder[A, ImmArray[A]] =
+    new mutable.Builder[A, ImmArray[A]] {
+      private[this] var size = 0
+      private[this] var array = new mutable.ArraySeq[A](initialCapacity max 1)
+
+      override def +=(elem: A): this.type = {
+        if (size == array.length) {
+          val newArray = new mutable.ArraySeq[A](size * 2)
+          copyArraySeq(array, 0, newArray, 0, size)
+          array = newArray
+        }
+        array(size) = elem
+        size += 1
+        this
+      }
+
+      override def clear(): Unit = size = 0
+
+      override def result(): ImmArray[A] =
+        if (size == 0) {
+          empty
+        } else {
+          val result = new ImmArray[A](0, size, array)
+          array = null
+          result
+        }
+    }
```

Comment:
I wouldn’t do this on the builder. Provide a `copy` function on the resulting `ImmArray`. The same thing doesn’t just apply to builders but also to slicing.

Reply:
Yeah, I know. However, with `copy` method in the `ImmArray` there's a slightly bigger chance of shooting yourself in the foot regarding memory usage if you hold on to the `ImmArray` using the full buffer anyway. For the builder, we know that we're the only owner of the buffer and hence copying is _always_ beneficial in terms of memory consumption. If there was just a language where you could express such lifetime and ownership considerations at the type level... :trollface:

Reply:
Adding a trimming at this point seems to have a noticeable impact on performance (without trimming: 31.769,  with trimming: 33.506, without optim:42.076).

I assume the performance impact comes mainly from the `CanBuildFrom` magic that uses a builder with default capacity. 
We probably can get back the good performance by avoiding at most as possible `ImmArraySeq` and `to[ImmArray]` in the critical part of our stack (Speedy, proto decoder). 

I will then add the trimming step.

Reply:
I don't think it's necessary that we trim during interpretation because most of the `ImmArray`s we create there are short-lived. I'm more worried about the `ImmArray`s we store in ASTs we hold onto for a long time.